main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala

```diff
@@ -226,28 +233,36 @@ object ConstraintSolver {
 
       resolveOneOf(curr, subst, renv) match {
         // Case 1: Success. Update the tracking variables.
-        case Result.Ok(ResolutionResult(newSubst, newConstrs, newProgress)) =>
+        case Result.Ok(IntermediateResult(newSubst, newConstrs, moreEffs, moreTraits, newProgress)) =>
           curr = newConstrs
           subst = newSubst
+          effs = moreEffs ::: effs
+          traits = moreTraits ::: traits
           progress = newProgress
 
         // Case 2: Error. Break out of the loop and return the error.
-        case res@Result.Err(_) =>
+        case Result.Err(err) =>
           Debug.stopRecording()
-          return res
+          return Result.Err(err)
       }
     }
-    Result.Ok(ResolutionResult(subst, curr, progress = true))
+
+    // After we reduced the normal constraints as far as possible, we try the effect constraints and the trait constraints.
+    for {
```

Comment:
I don't like that this is outside the while-loop. While this may not be the source of our current issues, this cannot in general work.

Reply:
We should really try to follow a model of always working on a constraint set. We are only done when it is empty or there is a conflict. There should be no other criteria/control-flow.

Reply:
Can I do this in a next revision?

I'd like to try to refactor by making a TypeConstraintSet which separates type constraints, effect constraints, and class constraints. I think that will help simplify things and make it easier to keep it all in the same loop.

Reply:
Yes. Sounds good on both accounts.